### DISCUSSION TITLE
[Feature] Time Travel for Your Memories: Date Photos by Age Recognition

### DISCUSSION THREAD

**Discussion by aviv926** [immich-app/immich#5133] [Feature Request]
### The feature

Many of us have old photos that we sent to be scanned by a computer or that we took ourselves, because these photos do not contain EXIF data, a situation arises that many of the photos do not appear in the timeline correctly, i.e. according to age
Therefore, after a little inquiry and research, I found out that Immich uses a face recognition model of the insightface project, where there is also the possibility of age recognition based on a photo.
An [example of such a code can be found here](https://github.com/deepinsight/insightface/blob/c2db41402c627cab8ea32d55da591940f2258276/attribute/gender_age/test.py) and because currently all Immich machine learning models run on a CPU, it is possible to set the code to use the CPU instead of GPU [as explained here](https://github.com/deepinsight/insightface/issues/619)

**Is there an onnx model?**
[Yes.](https://github.com/deepinsight/insightface/blob/c2db41402c627cab8ea32d55da591940f2258276/model_zoo/README.md#4-face-attribute-models
) And in fact the model buffalo_s up to buffalo_l including antelopev2 [support this feature](https://github.com/deepinsight/insightface/blob/c2db41402c627cab8ea32d55da591940f2258276/python-package/README.md#model-zoo)

**Why do I think this might be a good idea?**
Immich allows the addition of a person's birth date.
In other words, if a person was born in 2000, they are currently 23 years old. Therefore, in photos without original date data from EXIF, the age recognition model can provide an estimate of the age based on the photo and add an estimated date to the photo so that the location in the timeline is better.

**Limitations:** 
As of today,[ Immich cannot change information in external libraries due to read-only mode](https://immich.app/docs/features/libraries#:~:text=Immich%20will%20only%20read%20data%20from%20the%20files%2C%20and%20will%20not%20modify%20them%20in%20any%20way.)

**Idea for solving the limitations:**
option to let the admin choose whether to allow Immich to make changes to the images without Exif. There is no real reason not to allow such a change given the work that something is better than nothing

**Illustration of the process**
![3](https://github.com/immich-app/immich/assets/51673860/bcded6b9-dd78-496f-828b-edacf24977f0)

**Let me know what you think about it** 😊

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by mertalev** [comment]
This is an interesting idea, but I don't think the age recognition accuracy is high enough for this to be particularly useful. 

The [model zoo](https://github.com/deepinsight/insightface/tree/master/model_zoo) only reports the AgeDB-30 results for this task. AgeDB-30 is composed of pairs of people with an age difference of 30, testing the accuracy with which the model reports this difference (instead of 5, 10, or 20). 

Assuming the "genderage" model was trained on AgeDB, I expect it to be roughly accurate within +/- 5 years at best, at which point it's questionable whether the result is meaningful. It can also lead to images becoming out of order because the model inconsistently misidentifies the person's age across images. I don't expect this to significantly change in the future, given that it's difficult even for a human to be consistently accurate in estimating age closer than a few years.

Considering these limitations, I don't think this is worth the implementation effort.

**Reply by aviv926** [reply]
After much testing I was able to find a model called MiVOLO and it is incredibly accurate
There is even a [demo version in huggingface](https://huggingface.co/spaces/iitolstykh/age_gender_estimation_demo) and you can see that it has been trained on lots of datasets

Try yourself to perform tests on huggingface, I'm sure you'll be pleasantly surprised

**Here is a brief description:**
The most accurate and reliable age-classifier model today is the **MiVOLO: Multi-input Transformer for Age and Gender Estimation**¹. This model was developed by Maksim Kuprashevich and Irina Tolstykh from the Layer Team at SaluteDevices¹.

MiVOLO is a straightforward approach for age and gender estimation using the latest vision transformer¹. It integrates both tasks into a unified dual input/output model, leveraging not only facial information but also person image data¹. This improves the generalization ability of the model and enables it to deliver satisfactory results even when the face is not visible in the image¹.

The model was evaluated on four popular benchmarks and achieved state-of-the-art performance, while demonstrating real-time processing capabilities¹. It significantly outperforms humans across a majority of age ranges¹. The source code and data for this model can be accessed at their [GitHub repository](https://github.com/WildChlamydia/MiVOLO) 

Sources:
(1) MiVOLO: Multi-input Transformer for Age and Gender Estimation - arXiv.org. https://arxiv.org/pdf/2307.04616v1.pdf.
(2) MiVOLO: Multi-input Transformer for Age and Gender Estimation - arXiv.org. https://arxiv.org/pdf/2307.04616v1.pdf.
(3) 10 Best Machine Learning Algorithms (2023) - Unite.AI. https://www.unite.ai/ten-best-machine-learning-algorithms/.
(4) Accurate age classification using manual method and deep ... - Springer. https://link.springer.com/article/10.1007/s00414-021-02542-x.
(5) Designing an Age Classification Model with Deep Learning. https://heartbeat.comet.ml/designing-age-classifier-model-with-deep-learning-a346028d0530.

**Reply by eygraber** [reply]
I tried v1 and v2 with a bunch of photos and got results that were +/- 5 years on average.

**Reply by aviv926** [reply]
I decided to think of a different work method than machine learning for age prediction.
Take a look here:
https://github.com/immich-app/immich/pull/6037

**Comment by aviv926** [comment]
Rejected

[More info in discord](https://discord.com/channels/979116623879368755/1175956544312061992)
